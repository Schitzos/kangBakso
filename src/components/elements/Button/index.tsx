import theme from '@/styles/theme';
import React, { SVGProps } from 'react';
import { TouchableOpacity, Text, View, ViewStyle } from 'react-native';
import {styles} from './styles';

// Define the button sizes as a union type
type ButtonSize = 'small' | 'medium' | 'large';
type DimensionValue = ViewStyle['width'];

interface CustomButtonProps {
  onPress: () => void;
  size?: ButtonSize;
  buttonColor?: string;
  labelColor?: string;
  label: string;
  width?: DimensionValue; // Accept width as number or specific string values
  leftIcon?: React.ComponentType<SVGProps<SVGSVGElement>>;
  rightIcon?: React.ComponentType<SVGProps<SVGSVGElement>>;
  style?: ViewStyle;
  fontWeight?: 'normal' | 'bold' | '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900'; // Add fontWeight prop
  disabled?: boolean;
}

/**
 * A customizable button component with features such as:
 * - size (small, medium, large)
 * - button color
 * - label color
 * - label text
 * - width (accepts number or specific string values)
 * - left icon
 * - right icon
 * - custom styles
 * - font weight
 * - disabled property
 *
 * @param {{onPress: () => void, size?: ButtonSize, buttonColor?: string, labelColor?: string, label: string, width?: DimensionValue, leftIcon?: React.ComponentType<SVGProps<SVGSVGElement>>, rightIcon?: React.ComponentType<SVGProps<SVGSVGElement>>, style?: ViewStyle, fontWeight?: 'normal' | 'bold' | '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900', disabled?: boolean}} props
 * @returns {React.ReactElement} A React element representing the button
 */
const Button: React.FC<CustomButtonProps> = ({
  onPress,
  size = 'small',
  buttonColor = theme.colors.primary,
  labelColor = theme.colors.white,
  label,
  width = '100%', // Default width is full width
  leftIcon: LeftIcon,
  rightIcon: RightIcon,
  style,
  fontWeight = 'bold', // Set a default font weight
  disabled = false,
}) => {
  // Set styles based on size prop
  const sizeStyles = {
    small: styles.smallButton,
    medium: styles.mediumButton,
    large: styles.largeButton,
  };

  // Define height based on the size prop
  const heightStyles = {
    small: { height: 24 },
    medium: { height: 32 },
    large: { height: 40 },
  };

  // Define font size based on the size prop
  const fontSizeStyles = {
    small: { fontSize: 11 },
    medium: { fontSize: 14 },
    large: { fontSize: 14 },
  };
  return (
    <TouchableOpacity
      style={[
        styles.button,
        sizeStyles[size],
        heightStyles[size], // Apply height based on size
        { backgroundColor: buttonColor, width }, // Apply the width prop
        style,
      ]}
      onPress={onPress}
      disabled={disabled}
    >
      <View style={styles.content}>
        {LeftIcon && <LeftIcon width={16} height={16} />}
        <Text style={[styles.label, { color: labelColor, fontWeight }, fontSizeStyles[size]]}>
          {label}
        </Text>
        {RightIcon && <RightIcon width={16} height={16} />}
      </View>
    </TouchableOpacity>
  );
};

export default Button;
