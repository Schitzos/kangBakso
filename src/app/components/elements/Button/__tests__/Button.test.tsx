import React from 'react';
import { render } from '@testing-library/react-native';
import Button from '../index';

describe('Button Component', () => {
  test('matches snapshot with default props', () => {
    const { toJSON } = render(<Button label="Test" onPress={() => null}/>);
    expect(toJSON()).toMatchSnapshot();
  });
});
