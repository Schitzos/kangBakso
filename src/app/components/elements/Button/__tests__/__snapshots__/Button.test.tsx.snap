// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ED0226",
      "borderRadius": 56,
      "height": 24,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 10,
      "paddingVertical": 0,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 4,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Poppins-Regular",
          },
          {
            "color": "#FFFFFF",
            "fontWeight": "bold",
          },
          {
            "fontSize": 11,
          },
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;
